[roku] Roku Binding initial contribution

sdwilsh opened the pull request:
Initial, basic implementation of a binding for Roku devices.

This is #1338 with review comments addressed.

sdwilsh commented:
I debated prefixing this with `[WIP]`.  I think it's probably good enough as a starting point, and future work can go into follow-up PRs.

I'm willing to step up and implement more features for this binding as time goes on.

@kaikreuzer, I'm pretty sure this addresses all of your comments from #1338, but I haven't fully tested it yet (and I don't think I've looked at all the code in detail yet either).  I suspect you'll have more review comments, but if you don't, please don't merge this yet as I'd like to do some more testing and looking over the code :)

sdwilsh commented:
I forgot to add the `about.html` and the `README.md` :D

martinvw commented on addons/binding/org.openhab.binding.roku/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+Bundle-Name: Roku Binding
+Bundle-SymbolicName: org.openhab.binding.roku;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.1.0.qualifier
```
2.2.0.qualifier

martinvw commented on addons/binding/org.openhab.binding.roku/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+ org.slf4j
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.roku,
+ org.openhab.binding.roku.internal.handler
```
You should not export your internal's, either they should not be internal or not exported

martinvw commented on addons/binding/org.openhab.binding.roku/pom.xml:
```diff
@@ -0,0 +1,18 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.1.0-SNAPSHOT</version>
```
Can you please rebase and use version 2.2.0-SNAPSHOT

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/RokuBindingConstants.java:
```diff
@@ -0,0 +1,65 @@
+    public static final String BINDING_ID = "roku";
+
+    // List of all Thing Type UIDs
+    public final static ThingTypeUID THING_TYPE_ROKU = new ThingTypeUID(BINDING_ID, "roku");
```
Non-preferred order of modifiers (It was like this in a template, which has been changed some time ago) correct order is `public static final` See also: http://stackoverflow.com/questions/16731240/what-is-a-reasonable-order-of-java-modifiers-abstract-final-public-static-e#16732059

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/discovery/RokuDiscoveryParticipant.java:
```diff
@@ -0,0 +1,85 @@
+        properties.put(RokuBindingConstants.PORT, Integer.toString(port));
+        properties.put(RokuBindingConstants.REFRESH_INTERVAL, "30");
+        String label = "Roku Device";
+        if (state.getUserDeviceName() != null && !state.getUserDeviceName().equals("")) {
```
Position literals first in String comparisons - that way if the String is null you won't get a NullPointerException, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

sdwilsh replied:
Does this project run PMD automatically?

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/discovery/RokuDiscoveryParticipant.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    public ThingUID getThingUID(RemoteDevice device) {
```
Maybe name this method different, the sides effects in this method are imho not expected when reading this methods name.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/RokuState.java:
```diff
@@ -0,0 +1,120 @@
+ */
+public class RokuState {
+    public String udn;
+    public String serial_number;
```
Can you follow the naming conventions maybe with some annotation to map the fieldname?

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,125 @@
+            } catch (IOException e) {
+                logger.error("IOException occurred when attempting action", e);
+            }
+            updateState(channelUID, OnOffType.OFF);
```
Is OFF always the expected the response here, could you explain?

sdwilsh replied:
I've added a comment to clarify this.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,125 @@
+         * Future Functionality
+         * updateState(CHANNEL_APPBROWSER, state.getApplicationMenu());
+         */
+        if (!state.getPowerMode().equals("PowerOn")) {
```
Position literals first in String comparisons - that way if the String is null you won't get a NullPointerException, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,125 @@
+         * updateState(CHANNEL_APPBROWSER, state.getApplicationMenu());
+         */
+        if (!state.getPowerMode().equals("PowerOn")) {
+            updateStatus(ThingStatus.OFFLINE);
```
Maybe add a message that it is just OFF?

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,125 @@
+        properties.put(PROPERTY_MODEL_NAME, state.getModelName());
+        properties.put(Thing.PROPERTY_MODEL_ID, state.getModelNumber());
+        properties.put(PROPERTY_MODEL_REGION, state.getModelRegion());
+        if (state.getNetworkType().equals("ethernet")) {
```
Position literals first in String comparisons - that way if the String is null you won't get a NullPointerException, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

Please check all

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,125 @@
+        } else {
+            properties.put(PROPERTY_MAC, state.getWifiMac());
+        }
+        if (state.getUserDeviceName() != null && !state.getUserDeviceName().equals("")) {
```
If you want this check consider using `StringUtils.isNotEmpty()` just search for it in our code, we have apache commons 2.6 available by default.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommands.java:
```diff
@@ -0,0 +1,86 @@
+
+    private final Logger logger = LoggerFactory.getLogger(RokuCommands.class);
+
+    private RokuNetworkCalls rnc;
```
Consider making this final, since you assign it via the constructor, so always just once

sdwilsh replied:
I opted to delete this class instead.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+
+    /**
+     * Future Functionality
+     * private HashMap<String, String> getAllSubTags(String tag, Element eElement) {
```
Please remove this commented out code.

_if code is refactored in the meantime its unlikely that the commented out code is also modified, so as soon as the code might be needed it again it might not even work anymore_

_http://softwareengineering.stackexchange.com/questions/190096/can-commented-out-code-be-valuable-documentation_

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+            NodeList nList = doc.getElementsByTagName("device-info");
+            for (int i = 0; i < nList.getLength(); i++) {
+                Node nNode = nList.item(i);
+                logger.debug("Current Element: " + nNode.getNodeName());
```
You should be logging using placeholders, please check all!

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+                    }
+                } catch (NullPointerException e) {
+                    state.active_app = getTagName("app", eElement);
+                    String app_value = getSubTagName("app", eElement);
```
Please follow naming conventions!

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+                    String app_value = getSubTagName("app", eElement);
+                    try {
+                        state.active_app_img = new RawType(getData("/query/icon/" + app_value));
+                    } catch (Exception e1) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+                throw new IOException("Could not handle response");
+            }
+            return doc;
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

Please also note that you catch the exception thrown on line 168 in this catch...

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuNetworkCalls.java:
```diff
@@ -0,0 +1,77 @@
+
+    public String getResult(String content, String method) throws IOException {
+        StringBuilder result = new StringBuilder();
+        URL url = new URL("http://" + baseUrl + content);
```
This one has some potential resource leaks

martinvw commented:
Also travis / the static code analysis do report some errors (of which some are duplicate):

```
[ERROR] Failed to execute goal org.openhab.tools:static-code-analysis:0.0.6:report (default) on project org.openhab.binding.roku: Code Analysis Tool has found 7 error(s)!
[ERROR] Please fix the errors and rerun the build.
[ERROR] Errors list:
[ERROR] ERROR found by checkstyle: .binding.roku/META-INF/MANIFEST.MF:22 Remove internal package export org.openhab.binding.roku.internal.handler
[ERROR] ERROR found by checkstyle: .binding.roku/META-INF/MANIFEST.MF:28 The MANIFEST.MF file must not contain any Require-Bundle entries. Instead, Import-Package must be used.
[ERROR] ERROR found by findbugs: org.openhab.binding.roku.internal.protocol.RokuCommands.java:35 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.roku.internal.protocol.RokuCommunication.java:96 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.roku.internal.protocol.RokuCommunication.java:120 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.roku.internal.protocol.RokuNetworkCalls.java:39 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.roku.internal.protocol.RokuNetworkCalls.java:57 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] Detailed report can be found at: file:////home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.roku/target/code-analysis/report.html
[ERROR] -> [Help 1]
```

alanvanwyk commented:
Any news gents? 
Running openHAB 2.2.0 Build #1029 - but keen to test.

sdwilsh commented:
I should have some time to work on this later this month.  I just moved, so life in the non-digital world is a bit hectic.  Thanks for all the comments, @martinvw!

alanvanwyk commented:
Thanks Shawn,

Holler if you need testing - I've got a couple Rokus knocking about

On 7 Sep 2017, 16:53 +0100, Shawn Wilsher <notifications@github.com>, wrote:
> I should have some time to work on this later this month. I just moved, so life in the non-digital world is a bit hectic. Thanks for all the comments, @martinvw!
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

sdwilsh commented:
As I start to wind back up on this, @alanvanwyk, what kind of uses did you have in mind with this binding so I can make sure it's useful to more than just me :)

udo1toni commented:
@alanvanwyk I've got 2 rokus long time ago (in fact these are pinnacles) but to be honest, I only use them in squeezebox compability mode :)

alanvanwyk commented:
Hi guys,

I am creating a simple habpanel to share when I get visitors, which enables control of my AV gear and lighting in my house. It is coded in such way that the user can change AV source but the remote keeps the same look and feel (like a phone based universal remote, that works on any phone and only required a url, not an app install)

For now, I’ve just gone with html and it works OK. Having looked at it, the “API” (as you’re aware, it’s not a true API), is ok, but not really consistent (I guess to allow easy adaptation by app vendors)

Anyway, I’d prefer using proper managed bindings where possible, but have gotten around my current issue.




On 25 Sep 2017, 05:45 +0100, Udo Hartmann , wrote:
> @alanvanwyk I've got 2 rokus long time ago (in fact these are pinnacles) but to be honest, I only use them in squeezebox compability mode :)
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

sdwilsh commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,209 @@
+                            field.set(state, getTagName(methodStringArray[ii], eElement));
+                        } catch (NoSuchFieldException | SecurityException | IllegalArgumentException
+                                | IllegalAccessException e) {
+                            logger.error("Method not found {}", methodStringArray[ii].replace("-", "_"));
```
This was a highly deceptive logging message, and it should be `debug` level.

sdwilsh commented:
This isn't ready for another round of code review.  I have it at least updating with the Roku state again, but I need to do some more testing and address a few more review comments.  That will probably happen within the next week or two.

martinvw commented:
@sdwilsh any updates?

sdwilsh commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/config/RokuDeviceConfig.java:
```diff
@@ -0,0 +1,47 @@
+    public Thing getThing() {
+        return thing;
+    }
+}
```
It'd be nice if this were auto-formatted as desired.

sdwilsh commented:
Alright, @martinvw, I've addressed all the comments, and am reasonably comfortable with the implementation now.  Emulating the Roku remote is a bit weird to push into an openhab binding, but it works okay.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,157 @@
+            logger.debug("Current Element: {}", nNode.getNodeName());
+            try {
+                state.activeApp = getTagName("screensaver", eElement);
+                String app_value = getSubTagName("screensaver", eElement);
```
Also note that `app_value` does not follow the coding guidelines.

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/protocol/RokuCommunication.java:
```diff
@@ -0,0 +1,157 @@
+                try {
+                    state.activeAppImg = HttpUtil
+                            .downloadImage("http://" + host + ":" + port + "/query/icon/" + app_value);
+                } catch (Exception e) {
```
I'm not happy with this nested try-try-cath&retry-try-catch

martinvw commented on addons/binding/org.openhab.binding.roku/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+Bundle-Name: Roku Binding
+Bundle-SymbolicName: org.openhab.binding.roku;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
```
Can you update to 2.4.0

martinvw commented on addons/binding/org.openhab.binding.roku/OSGI-INF/RokuDiscovery.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2010-2017 by the respective copyright holders.
```
Some times has passed by can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

martinvw commented on addons/binding/org.openhab.binding.roku/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+ org.jupnp.model.meta,
+ org.openhab.binding.roku,
+ org.slf4j
+Service-Component: OSGI-INF/*.xml
```
We switched to using annotations to inject stuff, it would be great if you could adapt your code as well

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,131 @@
+    public void initialize() {
+        logger.debug("Initializing RokuDevice handler '{}'", getThing().getUID());
+        Number refreshInterval = (Number) this.getConfig().get(REFRESH_INTERVAL);
+        scheduler.scheduleWithFixedDelay(new Runnable() {
```
Please use a method reference ``this::updateData` should be able to replace the complete `Runnable`

martinvw commented on addons/binding/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:
```diff
@@ -0,0 +1,131 @@
+         * Future Functionality
+         * updateState(CHANNEL_APPBROWSER, state.getApplicationMenu());
+         */
+        if (!"PowerOn".equals(state.getPowerMode())) {
```
If you reverse the if you can get rid of the the `not` in general that reads better

jaywiseman1971 commented:
If you are looking for someone to test this; I have 12 Roku's in the house that range from Roku Stick, Roku 3 and Roku 4.  I'm running OH2 on a Synology NAS.

grimstoner commented:
I'm also keen to test.

meingraham commented:
@sdwilsh Shawn,

Any plans to complete this binding? Add me to the list of folks willing to test the binding. I have three Rokus (Roku 1 and 2 Roku LT).

If the effort to complete this and continue to support it has gone beyond the time you have available to devote to it, would you be willing to release the development to someone else? I'm not volunteering, I don't have the specific programming skills required. But if you are willing to let someone else take it from here, perhaps there are others who can see this to the finish line. I know that @rkoshak Rich Koshak had considered developing a Roku binding in late 2017/early 2018. That's one. I'd venture to guess that there are many others qualified and willing as well.

And, if that happens, my offer to help test stands, of course.

Regards.

Mike

sdwilsh commented:
I don't foresee myself having time to work on this for the next month or two.  I took over the original pull request, so I don't have an issue with someone else picking up where I left off (I already got permission from the initial pull request as well).  All my commits have been signed off already, so the code is there for someone to take.

jacksteraz commented:
@sdwilsh Any chance you could upload a jar of your binding so we could try it out?

J-N-K commented:
Closing due to inactivity. If someone wants to pick up this work, please open a new PR.
